src/tools/compiletest/src/read2.rs

```diff
@@ -5,29 +5,53 @@ pub use self::imp::read2;
 use std::io;
 use std::process::{Child, Output};
 
-pub fn read2_abbreviated(mut child: Child) -> io::Result<Output> {
+pub fn read2_abbreviated(mut child: Child, exclude_from_len: &[String]) -> io::Result<Output> {
     use io::Write;
     use std::mem::replace;
 
     const HEAD_LEN: usize = 160 * 1024;
     const TAIL_LEN: usize = 256 * 1024;
+    const EXCLUDED_PLACEHOLDER_LEN: isize = 32;
 
     enum ProcOutput {
-        Full(Vec<u8>),
+        Full { bytes: Vec<u8>, excluded_len: isize },
         Abbreviated { head: Vec<u8>, skipped: usize, tail: Box<[u8]> },
     }
 
     impl ProcOutput {
-        fn extend(&mut self, data: &[u8]) {
+        fn extend(&mut self, data: &[u8], exclude_from_len: &[String]) {
             let new_self = match *self {
-                ProcOutput::Full(ref mut bytes) => {
+                ProcOutput::Full { ref mut bytes, ref mut excluded_len } => {
                     bytes.extend_from_slice(data);
+
+                    // We had problems in the past with tests failing only in some environments,
+                    // due to the length of the base path pushing the output size over the limit.
+                    //
+                    // To make those failures deterministic across all environments we ignore known
+                    // paths when calculating the string length, while still including the full
+                    // path in the output. This could result in some output being larger than the
+                    // threshold, but it's better than having nondeterministic failures.
+                    //
+                    // The compiler emitting only excluded strings is addressed by adding a
+                    // placeholder size for each excluded segment, which will eventually reach
+                    // the configured threshold.
+                    for pattern in exclude_from_len {
+                        let pattern_bytes = pattern.as_bytes();
+                        let matches = data
+                            .windows(pattern_bytes.len())
+                            .filter(|window| window == &pattern_bytes)
+                            .count();
+                        *excluded_len += matches as isize
+                            * (EXCLUDED_PLACEHOLDER_LEN - pattern_bytes.len() as isize);
```

Comment:
This approach won't really work, I think, at least not in general. The data we receive here is an ~arbitrary slice of the total test output, which means that there's no guarantee that the pattern we're looking for isn't lying across a boundary (i.e., half the pattern in one "chunk" and half in the next).

I think the way to mitigate this would be to start matching here pattern_bytes-1 into the already pushed data, but I'm not entirely confident on the full correctness of that approach. (It may also be worth considering a RegexSet here, from a performance perspective, to do a single pass over the incoming bytes. But with our hard upper bound being pretty small it's probably not worth it).